hack/gen-docs.sh

```diff
@@ -0,0 +1,127 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# gen-docs.sh generates a new versioned docs directory under site/docs. It follows
+# the following process:
+#	1. Copies the contents of the most recently tagged docs directory into the new
+#	   directory, to establish a useful baseline to diff against.
+#	2. Adds all copied content from step 1 to git's staging area via 'git add'.
+#	3. Replaces the contents of the new docs directory with the contents of the
+#	   'master' docs directory, updating any version-specific links (e.g. to a
+#	   specific branch of the GitHub repository) to use the new version
+#   4. Copies the previous version's ToC file and runs 'git add' to establish
+#      a useful baseline to diff against.
+#   5. Replaces the content of the new ToC file with the master ToC.
+#   6. Update site/_config.yml and site/_data/toc-mapping.yml to include entries
+#      for the new version.
+#
+# The unstaged changes in the working directory can now easily be diff'ed against the
+# staged changes using 'git diff' to review all docs changes made since the previous 
+# tagged version. Once the unstaged changes are ready, they can be added to the
+# staging area using 'git add' and then committed.
+#
+# **NOTE**: there are additional manual steps required to finalize the process of generating
+# a new versioned docs site. The full process is documented in site/README-JEKYLL.md.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+# don't run if there's already a directory for the target docs version
+if [[ -d site/docs/$NEW_DOCS_VERSION ]]; then
+    echo "ERROR: site/docs/$NEW_DOCS_VERSION already exists"
+    exit 1
+fi
+
+# get the alphabetically last item in site/docs to use as PREVIOUS_DOCS_VERSION
+# if not explicitly specified by the user
+if [[ -z "${PREVIOUS_DOCS_VERSION:-}" ]]; then
+    echo "PREVIOUS_DOCS_VERSION was not specified, getting the latest version"
+    PREVIOUS_DOCS_VERSION=$(ls -1 site/docs/ | tail -n 1)
+fi
+
+# make a copy of the previous versioned docs dir
+echo "Creating copy of docs directory site/docs/$PREVIOUS_DOCS_VERSION in site/docs/$NEW_DOCS_VERSION"
+cp -r site/docs/${PREVIOUS_DOCS_VERSION}/ site/docs/${NEW_DOCS_VERSION}/
+
+# 'git add' the previous version's docs as-is so we get a useful diff when we copy the master docs in
+echo "Running 'git add' for previous version's doc contents to use as a base for diff"
+git add site/docs/${NEW_DOCS_VERSION}
+
+# now copy the contents of site/docs/master into the same directory so we can get a nice
+# git diff of what changed since previous version
+echo "Copying site/docs/master/ to site/docs/${NEW_DOCS_VERSION}/"
+rm -rf site/docs/${NEW_DOCS_VERSION}/ && cp -r site/docs/master/ site/docs/${NEW_DOCS_VERSION}/
+
+# make a copy of the previous versioned ToC
+NEW_DOCS_TOC="$(echo ${NEW_DOCS_VERSION} | tr . -)-toc"
```

Comment:
Wouldn't we want to copy `master`'s TOC here, rather than the previous version?

I'm imagining that we'd update `master` with any new ToC layouts, and carry that forward. Are you thinking that `master` and the current release would be updated in sync?

Reply:
This is to get a nice `git diff` also - so first copy over the previous version's ToC, `git add` it, then overwrite it with the `master` ToC so  you can now `git diff` to see what's changing since last version

Reply:
Oh, ok, so the same logic as the directory. Got it!